File: tests/algorithms/transpiler/FS/maths/special_numbers/harshad_numbers.fs
// Generated 2025-08-17 12:28 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let _repr v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", ", ")
let _floordiv (a:int) (b:int) : int =
    let q = a / b
    let r = a % b
    if r <> 0 && ((a < 0) <> (b < 0)) then q - 1 else q
let rec panic (msg: string) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable msg = msg
    try

        __ret
    with
        | Return -> __ret
and char_to_value (c: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable c = c
    try
        let digits: string = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        let mutable i: int = 0
        while i < (String.length (digits)) do
            if (string (digits.[i])) = c then
                __ret <- i
                raise Return
            i <- i + 1
        panic ("invalid digit")
        __ret
    with
        | Return -> __ret
and int_to_base (number: int) (``base``: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable number = number
    let mutable ``base`` = ``base``
    try
        if (``base`` < 2) || (``base`` > 36) then
            panic ("'base' must be between 2 and 36 inclusive")
        if number < 0 then
            panic ("number must be a positive integer")
        let digits: string = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        let mutable n: int = number
        let mutable result: string = ""
        while n > 0 do
            let remainder: int = ((n % ``base`` + ``base``) % ``base``)
            result <- (string (digits.[remainder])) + result
            n <- _floordiv (int n) (int ``base``)
        if result = "" then
            result <- "0"
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and base_to_int (num_str: string) (``base``: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable num_str = num_str
    let mutable ``base`` = ``base``
    try
        let mutable value: int = 0
        let mutable i: int = 0
        while i < (String.length (num_str)) do
            let c: string = string (num_str.[i])
            value <- (value * ``base``) + (char_to_value (c))
            i <- i + 1
        __ret <- value
        raise Return
        __ret
    with
        | Return -> __ret
and sum_of_digits (num: int) (``base``: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable num = num
    let mutable ``base`` = ``base``
    try
        if (``base`` < 2) || (``base`` > 36) then
            panic ("'base' must be between 2 and 36 inclusive")
        let num_str: string = int_to_base (num) (``base``)
        let mutable total: int = 0
        let mutable i: int = 0
        while i < (String.length (num_str)) do
            let c: string = string (num_str.[i])
            total <- total + (char_to_value (c))
            i <- i + 1
        __ret <- int_to_base (total) (``base``)
        raise Return
        __ret
    with
        | Return -> __ret
and harshad_numbers_in_base (limit: int) (``base``: int) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable limit = limit
    let mutable ``base`` = ``base``
    try
        if (``base`` < 2) || (``base`` > 36) then
            panic ("'base' must be between 2 and 36 inclusive")
        if limit < 0 then
            __ret <- Array.empty<string>
            raise Return
        let mutable numbers: string array = Array.empty<string>
        let mutable i: int = 1
        while i < limit do
            let s: string = sum_of_digits (i) (``base``)
            let divisor: int = base_to_int (s) (``base``)
            if (((i % divisor + divisor) % divisor)) = 0 then
                numbers <- Array.append numbers [|(int_to_base (i) (``base``))|]
            i <- i + 1
        __ret <- numbers
        raise Return
        __ret
    with
        | Return -> __ret
and is_harshad_number_in_base (num: int) (``base``: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable num = num
    let mutable ``base`` = ``base``
    try
        if (``base`` < 2) || (``base`` > 36) then
            panic ("'base' must be between 2 and 36 inclusive")
        if num < 0 then
            __ret <- false
            raise Return
        let mutable n: string = int_to_base (num) (``base``)
        let d: string = sum_of_digits (num) (``base``)
        let n_val: int = base_to_int (n) (``base``)
        let d_val: int = base_to_int (d) (``base``)
        __ret <- (((n_val % d_val + d_val) % d_val)) = 0
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        ignore (printfn "%s" (int_to_base (0) (21)))
        ignore (printfn "%s" (int_to_base (23) (2)))
        ignore (printfn "%s" (int_to_base (58) (5)))
        ignore (printfn "%s" (int_to_base (167) (16)))
        ignore (printfn "%s" (sum_of_digits (103) (12)))
        ignore (printfn "%s" (sum_of_digits (1275) (4)))
        ignore (printfn "%s" (sum_of_digits (6645) (2)))
        ignore (printfn "%s" (_repr (harshad_numbers_in_base (15) (2))))
        ignore (printfn "%s" (_repr (harshad_numbers_in_base (12) (34))))
        ignore (printfn "%s" (_repr (harshad_numbers_in_base (12) (4))))
        ignore (printfn "%b" (is_harshad_number_in_base (18) (10)))
        ignore (printfn "%b" (is_harshad_number_in_base (21) (10)))
        ignore (printfn "%b" (is_harshad_number_in_base (-21) (5)))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
ignore (main())

File: tests/algorithms/x/Pascal/maths/special_numbers/harshad_numbers.pas
{$mode objfpc}{$modeswitch nestedprocvars}
program Main;
uses SysUtils;
type StrArray = array of string;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
procedure error(msg: string);
begin
  panic(msg);
end;
function _to_float(x: integer): real;
begin
  _to_float := x;
end;
function to_float(x: integer): real;
begin
  to_float := _to_float(x);
end;
procedure json(xs: array of real);
var i: integer;
begin
  write('[');
  for i := 0 to High(xs) do begin
    write(xs[i]);
    if i < High(xs) then write(', ');
  end;
  writeln(']');
end;
function list_to_str(xs: array of string): string;
var i: integer;
begin
  Result := '#(' + sLineBreak;
  for i := 0 to High(xs) do begin
    Result := Result + '  ''' + xs[i] + '''.' + sLineBreak;
  end;
  Result := Result + ')';
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  base: integer;
  c: string;
  limit: integer;
  msg: string;
  num: integer;
  num_str: string;
  number: integer;
procedure panic_(msg: string); forward;
function char_to_value(c: string): integer; forward;
function int_to_base(number: integer; base: integer): string; forward;
function base_to_int(num_str: string; base: integer): integer; forward;
function sum_of_digits(num: integer; base: integer): string; forward;
function harshad_numbers_in_base(limit: integer; base: integer): StrArray; forward;
function is_harshad_number_in_base(num: integer; base: integer): boolean; forward;
procedure main(); forward;
procedure panic_(msg: string);
begin
end;
function char_to_value(c: string): integer;
var
  char_to_value_digits: string;
  char_to_value_i: integer;
begin
  char_to_value_digits := '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  char_to_value_i := 0;
  while char_to_value_i < Length(char_to_value_digits) do begin
  if char_to_value_digits[char_to_value_i+1] = c then begin
  exit(char_to_value_i);
end;
  char_to_value_i := char_to_value_i + 1;
end;
  panic_('invalid digit');
end;
function int_to_base(number: integer; base: integer): string;
var
  int_to_base_digits: string;
  int_to_base_n: integer;
  int_to_base_result_: string;
  int_to_base_remainder: integer;
begin
  if (base < 2) or (base > 36) then begin
  panic_('''base'' must be between 2 and 36 inclusive');
end;
  if number < 0 then begin
  panic_('number must be a positive integer');
end;
  int_to_base_digits := '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  int_to_base_n := number;
  int_to_base_result_ := '';
  while int_to_base_n > 0 do begin
  int_to_base_remainder := int_to_base_n mod base;
  int_to_base_result_ := int_to_base_digits[int_to_base_remainder+1] + int_to_base_result_;
  int_to_base_n := int_to_base_n div base;
end;
  if int_to_base_result_ = '' then begin
  int_to_base_result_ := '0';
end;
  exit(int_to_base_result_);
end;
function base_to_int(num_str: string; base: integer): integer;
var
  base_to_int_value: integer;
  base_to_int_i: integer;
  base_to_int_c: string;
begin
  base_to_int_value := 0;
  base_to_int_i := 0;
  while base_to_int_i < Length(num_str) do begin
  base_to_int_c := num_str[base_to_int_i+1];
  base_to_int_value := (base_to_int_value * base) + char_to_value(base_to_int_c);
  base_to_int_i := base_to_int_i + 1;
end;
  exit(base_to_int_value);
end;
function sum_of_digits(num: integer; base: integer): string;
var
  sum_of_digits_num_str: string;
  sum_of_digits_total: integer;
  sum_of_digits_i: integer;
  sum_of_digits_c: string;
begin
  if (base < 2) or (base > 36) then begin
  panic_('''base'' must be between 2 and 36 inclusive');
end;
  sum_of_digits_num_str := int_to_base(num, base);
  sum_of_digits_total := 0;
  sum_of_digits_i := 0;
  while sum_of_digits_i < Length(sum_of_digits_num_str) do begin
  sum_of_digits_c := sum_of_digits_num_str[sum_of_digits_i+1];
  sum_of_digits_total := sum_of_digits_total + char_to_value(sum_of_digits_c);
  sum_of_digits_i := sum_of_digits_i + 1;
end;
  exit(int_to_base(sum_of_digits_total, base));
end;
function harshad_numbers_in_base(limit: integer; base: integer): StrArray;
var
  harshad_numbers_in_base_numbers: array of string;
  harshad_numbers_in_base_i: integer;
  harshad_numbers_in_base_s: string;
  harshad_numbers_in_base_divisor: integer;
begin
  if (base < 2) or (base > 36) then begin
  panic_('''base'' must be between 2 and 36 inclusive');
end;
  if limit < 0 then begin
  exit([]);
end;
  harshad_numbers_in_base_numbers := [];
  harshad_numbers_in_base_i := 1;
  while harshad_numbers_in_base_i < limit do begin
  harshad_numbers_in_base_s := sum_of_digits(harshad_numbers_in_base_i, base);
  harshad_numbers_in_base_divisor := base_to_int(harshad_numbers_in_base_s, base);
  if (harshad_numbers_in_base_i mod harshad_numbers_in_base_divisor) = 0 then begin
  harshad_numbers_in_base_numbers := concat(harshad_numbers_in_base_numbers, StrArray([int_to_base(harshad_numbers_in_base_i, base)]));
end;
  harshad_numbers_in_base_i := harshad_numbers_in_base_i + 1;
end;
  exit(harshad_numbers_in_base_numbers);
end;
function is_harshad_number_in_base(num: integer; base: integer): boolean;
var
  is_harshad_number_in_base_n: string;
  is_harshad_number_in_base_d: string;
  is_harshad_number_in_base_n_val: integer;
  is_harshad_number_in_base_d_val: integer;
begin
  if (base < 2) or (base > 36) then begin
  panic_('''base'' must be between 2 and 36 inclusive');
end;
  if num < 0 then begin
  exit(false);
end;
  is_harshad_number_in_base_n := int_to_base(num, base);
  is_harshad_number_in_base_d := sum_of_digits(num, base);
  is_harshad_number_in_base_n_val := base_to_int(is_harshad_number_in_base_n, base);
  is_harshad_number_in_base_d_val := base_to_int(is_harshad_number_in_base_d, base);
  exit((is_harshad_number_in_base_n_val mod is_harshad_number_in_base_d_val) = 0);
end;
procedure main();
begin
  writeln(int_to_base(0, 21));
  writeln(int_to_base(23, 2));
  writeln(int_to_base(58, 5));
  writeln(int_to_base(167, 16));
  writeln(sum_of_digits(103, 12));
  writeln(sum_of_digits(1275, 4));
  writeln(sum_of_digits(6645, 2));
  writeln(list_to_str(harshad_numbers_in_base(15, 2)));
  writeln(list_to_str(harshad_numbers_in_base(12, 34)));
  writeln(list_to_str(harshad_numbers_in_base(12, 4)));
  writeln(Ord(is_harshad_number_in_base(18, 10)));
  writeln(Ord(is_harshad_number_in_base(21, 10)));
  writeln(Ord(is_harshad_number_in_base(-21, 5)));
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  main();
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
